### ISSUE TITLE
etcd.service can start and fail before /var/lib/etcddisk is mounted

### ISSUE THREAD

**Issue by UnwashedMeme** [Azure/aks-engine#1515]
**Describe the bug**

There is currently no ordering specified to etcd that forces the disk to be present before starting etcd. By and large this works correctly but when an Azure hypervisor crashes and the machine is redeployed and the disk needs to be fscked...

```
$ journalctl -b | grep -E 'etcddisk|sdc1'                                                                                                                    
Jun 20 07:51:26 k8s-master-38523805-0 kernel:  sdc: sdc1                                                                                                                                      
Jun 20 07:51:28 k8s-master-38523805-0 systemd[1]: Starting File System Check on /dev/sdc1...                                                                                                  
Jun 20 07:51:41 k8s-master-38523805-0 etcd[1115]: cannot access data directory: open /var/lib/etcddisk/.touch: permission denied                                                              
Jun 20 07:51:42 k8s-master-38523805-0 etcd[1616]: cannot access data directory: open /var/lib/etcddisk/.touch: permission denied                                                              
Jun 20 07:51:42 k8s-master-38523805-0 etcd[1659]: cannot access data directory: open /var/lib/etcddisk/.touch: permission denied                                                              
Jun 20 07:51:42 k8s-master-38523805-0 etcd[1712]: cannot access data directory: open /var/lib/etcddisk/.touch: permission denied                                                              
Jun 20 07:51:43 k8s-master-38523805-0 etcd[1786]: cannot access data directory: open /var/lib/etcddisk/.touch: permission denied                                                              
Jun 20 07:52:04 k8s-master-38523805-0 kubelet[1354]: I0620 07:52:04.698265    1354 fs.go:144] Filesystem UUIDs: map[1483fb68-35cf-482d-b933-9350a9c36988:/dev/sdc1 3fa8625b-9124-44c9-a434-7a6d
fda8d1b6:/dev/sdb1 D183-43BF:/dev/sda15 eb5f4234-4e60-46a9-a0dc-76fd1a593cdd:/dev/sda1]                                                                                                       
Jun 20 07:52:21 k8s-master-38523805-0 systemd[1]: Started File System Check on /dev/sdc1.                                                                                                     
Jun 20 07:52:21 k8s-master-38523805-0 audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj==unconfined msg='unit=systemd-fsck@dev-sdc1 comm="systemd" exe="/lib/systemd/syst
emd" hostname=? addr=? terminal=? res=success'                                                                                                                                                
Jun 20 07:52:21 k8s-master-38523805-0 kernel: audit: type=1130 audit(1561017141.303:171): pid=1 uid=0 auid=4294967295 ses=4294967295 subj==unconfined msg='unit=systemd-fsck@dev-sdc1 comm="sys
temd" exe="/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'                                                                                                                    
Jun 20 07:52:21 k8s-master-38523805-0 systemd[1]: Mounting /var/lib/etcddisk...                                                                                                               
Jun 20 07:52:21 k8s-master-38523805-0 kernel: EXT4-fs (sdc1): mounted filesystem with ordered data mode. Opts: (null)                                                                         
Jun 20 07:52:21 k8s-master-38523805-0 systemd[1]: Mounted /var/lib/etcddisk.                                                                                                                  
Jun 20 13:35:16 k8s-master-38523805-0 sudo[77207]: azureuser : TTY=pts/0 ; PWD=/home/azureuser ; USER=root ; COMMAND=/bin/ls -hal /var/lib/etcddisk  
```
(the last line is me logging in hours later going "wtf")

I think this is related to https://github.com/Azure/acs-engine/issues/1481, and there was an attempt to re-establish a chain requiring this ordering in https://github.com/Azure/acs-engine/pull/2774, but that wasn't complete and got lost when acs-engine->aks-engine.

**AKS Engine version**
* aks-engine: v0.35.1 (I think, can you find this on the master anywhere?)
* Master is not a vmss
* distro: `aks-18.04`

**Kubernetes version**
v1.14.1

**Additional context**
```dot
$  systemd-analyze dot etcd.service var-lib-etcddisk.mount
digraph systemd {
	"var-lib-etcddisk.mount"->"system.slice" [color="green"];
	"var-lib-etcddisk.mount"->"systemd-journald.socket" [color="green"];
	"var-lib-etcddisk.mount"->"-.mount" [color="green"];
	"var-lib-etcddisk.mount"->"systemd-fsck@dev-sdc1.service" [color="green"];
	"var-lib-etcddisk.mount"->"dev-sdc1.device" [color="green"];
	"var-lib-etcddisk.mount"->"local-fs-pre.target" [color="green"];
	"var-lib-etcddisk.mount"->"-.mount" [color="black"];
	"var-lib-etcddisk.mount"->"systemd-fsck@dev-sdc1.service" [color="black"];
	"var-lib-etcddisk.mount"->"system.slice" [color="black"];
	"var-lib-etcddisk.mount"->"umount.target" [color="red"];
	"dev-sdc1.device"->"var-lib-etcddisk.mount" [color="grey66"];
	"shutdown.target"->"etcd.service" [color="green"];
	"multi-user.target"->"etcd.service" [color="green"];
	"multi-user.target"->"etcd.service" [color="grey66"];
	"local-fs.target"->"var-lib-etcddisk.mount" [color="grey66"];
	"umount.target"->"var-lib-etcddisk.mount" [color="green"];
	"etcd.service"->"systemd-journald.socket" [color="green"];
	"etcd.service"->"network.target" [color="green"];
	"etcd.service"->"system.slice" [color="green"];
	"etcd.service"->"sysinit.target" [color="green"];
	"etcd.service"->"basic.target" [color="green"];
	"etcd.service"->"system.slice" [color="black"];
	"etcd.service"->"sysinit.target" [color="black"];
	"etcd.service"->"network-online.target" [color="grey66"];
	"etcd.service"->"shutdown.target" [color="red"];
}
```

**Comment by jackfrancis** [comment]
@UnwashedMeme Are we expecting that by moving the mount logic to cloud-init that such a reboot-after-hypervisor-failure scenario would improve? Here's when this was originally moved into CustomScriptExtension:

https://github.com/Azure/acs-engine/pull/661/files

The inverse of that is something like what it would be in cloud-init.

I'm not convinced that moving the mount statement to cloud-init would improve the scenario you describe; off the top of my head I'd lean towards thinking we'd need to use systemd to block on the mount reconciliation before launching the runtime.

Thoughts?

**Comment by UnwashedMeme** [comment]
@jackfrancis When I started digging on this I found https://github.com/Azure/aks-engine/blob/master/parts/k8s/cloud-init/artifacts/mountetcd.sh#L2  The issues I linked above I think were an attempt to undo this and provide some means to enforce ordering but I didn't dig very deeply.

My proposal is as you say, use systemd to handle this: add `RequiresMountsFor=/var/lib/etcddisk` to the `etcd.service` file. I think this is a useful bit regardless of customscript/cloudinit differences.

Before adding that:
![etcd](https://user-images.githubusercontent.com/51197/60199137-47eb9c00-9811-11e9-8903-914822dcafbd.png)

After adding that:
![etcd](https://user-images.githubusercontent.com/51197/60199238-77020d80-9811-11e9-92d6-2b32541f1863.png)

That's looking just right.

---

I'm trying to test this but can't reproduce well. Setting the max mount count on the disk (`tune2fs -c 1 /dev/sdc1`) appears to trigger a fsck but it finishes fast.
```
Jun 26 16:42:54 k8s-master-40037667-0 systemd[1]: Started File System Check on /dev/sdc1.
Jun 26 16:42:54 k8s-master-40037667-0 audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj==unconfined msg='unit=systemd-fsck@dev-sdc1 comm="systemd" exe="/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Jun 26 16:42:54 k8s-master-40037667-0 systemd[1]: Mounting /var/lib/etcddisk...
Jun 26 16:42:54 k8s-master-40037667-0 systemd[1]: Mounted /var/lib/etcddisk.
Jun 26 16:42:54 k8s-master-40037667-0 kernel: EXT4-fs (sdc1): mounted filesystem with ordered data mode. Opts: (null)
```
that's all within 1 second so it doesn't really delay startup like the original error where fsck appeared to take about 55seconds.

So systemd is reporting that the service depends on the mount, but I can't independently verify.
